000X-B-C-I-R-S-error-handling-codes.md

```diff
@@ -0,0 +1,337 @@
+  ...
+}
+```
+
+### Dapr Custom Error Info
+```go
+message DaprKitErrorInfo {
+  string reason = 1;
+  string domain = 2;
+  map<string, string> metadata = 3;
+}
+```
+
+### Example Dapr Custom Error Info
+```json
+{
+  "code": 8089,
+  "message": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused",
+  "details": [
+    {
+      "@type": "type.googleapis.com/google.rpc.ErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    },
+    {
+      "@type": "type.googleapis.com/google.rpc.ResourceInfo",
+      "resource_type": "StateStore",
+      "resource_name": "REDIS",
+      "owner": "",
+      "description": "unable to perform Redis.HGETALL call: dial tcp [::1]:6379: connect: connection refused"
+    },
+    {
+      "@type": "type.googleapis.com/kit.proto.customerrors.v1.DaprKitErrorInfo",
+      "reason": "DPR_COMP_STS_1",
+      "domain": "redis.state.contrib.dapr.io",
+      "metadata": {
+        "key": "myapp||name"
+      }
+    }
+  ]
+}
+```
+### Pros
+- Since the Dapr Runtime is using protocol buffers as the data format, support for the richer error model is already included in most of the gRPC implementations.
+- This would help minimize the changes with the Dapr ecosystem.
+- This solution could be used to programmatically react to errors as it provides a standard structure for the errors with details.
+
+### Cons
+- Dependencies on gPRC richer error model.
+- Not all gRPC implementations support this out of the box, [C#/dotnet](https://learn.microsoft.com/en-us/dotnet/architecture/grpc-for-wcf-developers/error-handling#grpc-richer-error-model) being one of them, but there is a [possible solution (Google.Api.CommonProtos)](https://www.nuget.org/packages/Google.Api.CommonProtos/) to this.
+
+
+## gRPC Richer Error Model POC
+For the POC I've made changes to some parts of the Dapr modules (). The POC code can be found in my GH Repo under the branch **error-codes-poc**
+
+These are the gRPC imports used:
+
+```go
+import (
+  ...
+  "google.golang.org/genproto/googleapis/rpc/errdetails"
+  "google.golang.org/grpc/codes"
+  "google.golang.org/grpc/status"
+  ...
+)
+```
+
+The files changed for this POC:
+
+https://github.com/robertojrojas/components-contrib/tree/error-codes-poc
+
+- state/redis/redis.go
+- state/store.go
+
+https://github.com/robertojrojas/dapr-kit/tree/error-codes-poc
+
+- pkg/proto/customerrors/v1/customerrors.pb.go
+- proto/customerrors/v1/customerrors.proto
+- status/customerrorcodes.go
+- status/status.go
+
+https://github.com/robertojrojas/dapr/tree/error-codes-poc
+
+- pkg/diagnostics/grpc_tracing.go
+- pkg/grpc/api.go
+- pkg/http/api.go
+- pkg/http/responses.go
+
+https://github.com/robertojrojas/dapr-go-sdk/tree/error-codes-poc
+
+- client/state.go
+
+https://github.com/robertojrojas/dapr-cli/tree/error-codes-poc
+
+- pkg/standalone/invoke.go
+
+
+
+## Solution B
```

Comment:
Mention this as alternative consideration but the proposal should have only one solution and not two.